utils/src/test/java/software/amazon/awssdk/utils/async/ByteBufferStoringSubscriberTest.java

```diff
@@ -0,0 +1,308 @@
+        assertThat(out.hasRemaining()).isFalse();
+    }
+
+    @Test
+    @Timeout(30)
+    public void stochastic_subscriberSeemsThreadSafe() throws Throwable {
+        ExecutorService producer = Executors.newFixedThreadPool(1);
+        ExecutorService consumer = Executors.newFixedThreadPool(1);
+        try {
+            ByteBufferStoringSubscriber subscriber = new ByteBufferStoringSubscriber(50);
+
+            AtomicBoolean testRunning = new AtomicBoolean(true);
+            AtomicInteger messageNumber = new AtomicInteger(0);
+
+            AtomicReference<Throwable> producerFailure = new AtomicReference<>();
+            Subscription subscription = new Subscription() {
+                @Override
+                public void request(long n) {
+                    producer.submit(() -> {
+                        try {
+                            for (int i = 0; i < n; i++) {
+                                ByteBuffer buffer = ByteBuffer.allocate(4);
+                                buffer.putInt(messageNumber.getAndIncrement());
+                                buffer.flip();
+                                subscriber.onNext(buffer);
+                            }
+                        } catch (Throwable t) {
+                            producerFailure.set(t);
+                        }
+                    });
+                }
+
+                @Override
+                public void cancel() {
+                    producerFailure.set(new AssertionError("Cancel not expected."));
+                }
+            };
+
+            subscriber.onSubscribe(subscription);
+
+            Future<Object> consumerFuture = consumer.submit(() -> {
+                ByteBuffer carryOver = ByteBuffer.allocate(4);
+
+                int expectedMessageNumber = 0;
+                while (testRunning.get()) {
+                    Thread.sleep(1);
+
+                    ByteBuffer out = ByteBuffer.allocate(4 + expectedMessageNumber);
+                    subscriber.transferTo(out);
+
+                    out.flip();
+
+                    if (carryOver.position() > 0) {
+                        int oldOutLimit = out.limit();
+                        out.limit(carryOver.remaining());
+                        carryOver.put(out);
+                        out.limit(oldOutLimit);
+
+                        carryOver.flip();
+                        assertThat(carryOver.getInt()).isEqualTo(expectedMessageNumber);
+                        ++expectedMessageNumber;
+                        carryOver.clear();
+                    }
+
+                    while (out.remaining() >= 4) {
+                        assertThat(out.getInt()).isEqualTo(expectedMessageNumber);
+                        ++expectedMessageNumber;
+                    }
+
+                    if (out.hasRemaining()) {
+                        carryOver.put(out);
+                    }
+                }
+                return null;
+            });
+
+            Thread.sleep(5_000);
```

Comment:
Will this test be flaky?

Reply:
The sleep duration is just how long to spend trying to break it. It's not manual synchronization to make things work. 10 ms would work, just be less reliable of a test.

If there's a bug, this test will be flaky. If there's not a bug, the test should never fail.

Reply:
I ran it for a few minutes and never saw it fail.